src/Features/CSharp/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.cs

```diff
@@ -0,0 +1,96 @@
+        }
+
+        protected override SyntaxNode GetPartialTypeSyntaxNode(SyntaxTree tree, int position, CancellationToken cancellationToken)
+        {
+            TypeDeclarationSyntax declaration;
+            return tree.IsPartialTypeDeclarationNameContext(position, cancellationToken, out declaration) ? declaration : null;
+        }
+
+        protected override Task<AbstractSyntaxContext> CreateSyntaxContextAsync(Document document, SemanticModel semanticModel, int position, CancellationToken cancellationToken)
+        {
+            return Task.FromResult<AbstractSyntaxContext>(CSharpSyntaxContext.CreateContext(document.Project.Solution.Workspace, semanticModel, position, cancellationToken));
+        }
+
+        protected override ValueTuple<string, string> GetDisplayAndInsertionText(INamedTypeSymbol symbol, AbstractSyntaxContext context)
+        {
+            var displayAndInsertionText = symbol.ToMinimalDisplayString(context.SemanticModel, context.Position, _symbolFormatWithGenerics);
+            return ValueTuple.Create(displayAndInsertionText, displayAndInsertionText);
+        }
+
+        protected override IEnumerable<INamedTypeSymbol> LookupCandidateSymbols(AbstractSyntaxContext context, INamedTypeSymbol declaredSymbol, CancellationToken cancellationToken)
+        {
+            var candidates = base.LookupCandidateSymbols(context, declaredSymbol, cancellationToken);
+
+            // The base class applies a broad filter when finding candidates, but since C# requires
+            // that all parts have the "partial" modifier, the results can be trimmed further here.
+            return candidates?.Where(symbol => symbol.DeclaringSyntaxReferences.Any(reference => IsPartialTypeDeclaration(reference.GetSyntax(cancellationToken))));
+        }
+
+        private static bool IsPartialTypeDeclaration(SyntaxNode syntax)
+        {
+            var declarationSyntax  = syntax as BaseTypeDeclarationSyntax;
+            return declarationSyntax != null && declarationSyntax.Modifiers.Any(modifier => modifier.IsKind(SyntaxKind.PartialKeyword));
+        }
+
+        public async override Task<TextChange?> GetTextChangeAsync(Document document, CompletionItem selectedItem, char? ch, CancellationToken cancellationToken)
+        {
+            if (ch == '<')
+            {
+                var symbols = await SymbolCompletionItem.GetSymbolsAsync(selectedItem, document, cancellationToken).ConfigureAwait(false);
+
+                if (symbols.Length > 0)
+                {
+                    var position = SymbolCompletionItem.GetContextPosition(selectedItem);
+                    var semanticModel = await document.GetSemanticModelForSpanAsync(new TextSpan(position, 0), cancellationToken).ConfigureAwait(false);
+                    var insertionText = symbols[0].ToMinimalDisplayString(semanticModel, position, _symbolFormatWithoutGenerics);
```

Comment:
> This also deals with escaping things like @class

That's pretty easily handled by just saying: symbol.Name.EscapeIdentifier() though.  SymbolDisplay is really a good choice when it's a long and complicated name (i.e. when there's a signature or generics in it).  But when there's literally just the name alone, it serves little purpose.